### PR TITLE
Add new properties for communication methods to user

### PR DESCRIPTION
**Pull request by borovHokage**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/112

### CODE REVIEW

File: src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs [hunk 1]
```diff
@@ -27,7 +28,14 @@ public DbUser Map(UserRequest value)
                 Status = value.Status,
                 AvatarFileId = value.AvatarFileId,
                 IsActive = value.IsActive,
-                IsAdmin = value.IsAdmin
+                IsAdmin = value.IsAdmin,
+                Connections = value.Connection?.Select(x => new DbConnection
```
Comment (MariyaSin) [inline src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs:32]: just a question: why is here  "?" ?

File: src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs [hunk 2]
```diff
@@ -27,7 +28,14 @@ public DbUser Map(UserRequest value)
...
+                {
+                    Id =Guid.NewGuid(),
+                    Type = (int)x.Type,
+                   Value = x.Value,
+                   UserId = value.Id.Value
```
Comment (MariyaSin) [inline src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs:37]: you need to make the same indents for properties

File: src/UserService.Models.Db/DbConnection.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    public class DbConnectionConfiguration : IEntityTypeConfiguration<DbConnection>
+    {
+        public void Configure(EntityTypeBuilder<DbConnection> builder)
+        {
+            builder.ToTable("UserConnections");
+            
+            builder.HasKey(pm => new {pm.UserId, pm.Id});
+            builder.HasOne(pm => pm.User)
+                .WithMany(pm => pm.Connections)
+                .HasForeignKey(pm => pm.UserId);
```
Comment (MariyaSin) [inline src/UserService.Models.Db/DbConnection.cs:30]: Why do you use a composite key?

File: src/UserService.Models.Dto/UserConnection.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.UserService.Models.Dto
+{
+    public class UserConnection
+    {
+        public ConnectionType  Type { get; set; }
+        public string Value { get; set; }
+
```
Comment (MariyaSin) [inline src/UserService.Models.Dto/UserConnection.cs:13]: remove empty line

File: src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs [hunk 5]
```diff
@@ -12,7 +13,7 @@ public DbUser Map(UserRequest value)
         {
             if (value == null)
             {
-                throw new BadRequestException();
+                throw new BadRequestException(); 
             }
 
             value.Id ??= Guid.NewGuid();
```
Comment (BelehovEgor) [inline src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs:19]: What does "??" mean?

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                         principalTable: "Users",
+                         principalColumn: "Id",
+                         onDelete: ReferentialAction.Cascade);
+             });
+        }
+        protected override void Down(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.DropTable(
+                name: "UserConnections");
+
```
[the remark this answers is not in the document]
Comment (borovHokage) [reply src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs]: Fixed

File: src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs (hunk 1 above)
Comment (borovHokage) [reply src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs:32]: Because here we are checking the conditions of the object(it shouldn't be null)

File: src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs (hunk 5 above)
Comment (borovHokage) [reply src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs:19]: Sorry, that isn't my part of the code. My changes are marked "+"

Comment (BelehovEgor) [reply src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs:19]: Oh, I see

File: src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs (hunk 2 above)
Comment (borovHokage) [reply src/UserService.Mappers/RequestsMappers/UserRequestMapper.cs]: fixed indents

File: src/UserService.Models.Dto/UserConnection.cs (hunk 4 above)
Comment (borovHokage) [reply src/UserService.Models.Dto/UserConnection.cs]: fixed

File: src/UserService.Models.Db/DbConnection.cs (hunk 3 above)
Comment (borovHokage) [reply src/UserService.Models.Db/DbConnection.cs:30]: to make a connection one to many. A lot of "Connection" to one "User"

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                 table.PrimaryKey("PK_UserConnection", a => a.Id);
+                 table.ForeignKey(
+                         name: "FK_UserConnections_Users",
+                         column: x => x.UserId,
+                         principalTable: "Users",
+                         principalColumn: "Id",
+                         onDelete: ReferentialAction.Cascade);
+             });
+        }
+        protected override void Down(MigrationBuilder migrationBuilder)
```
Comment (djino97) [inline src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs:35]: add empty line

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+                UserId = table.Column<Guid>(nullable: false)
+            },
+             constraints: table => 
+             {
+                 table.PrimaryKey("PK_UserConnection", a => a.Id);
+                 table.ForeignKey(
+                         name: "FK_UserConnections_Users",
+                         column: x => x.UserId,
+                         principalTable: "Users",
+                         principalColumn: "Id",
```
Comment (djino97) [inline src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs:31]: add property through nameof(DbUser.Id)

File: test/UserService.Mappers.UnitTests/RequestsMappers/UserRequestMapperTests.cs [hunk 9]
```diff
@@ -4,6 +4,8 @@
 using LT.DigitalOffice.UserService.Models.Dto;
 using NUnit.Framework;
 using System;
+using System.Collections.Generic;
+using System.Linq;
 
 namespace LT.DigitalOffice.UserService.Mappers.RequestsMappers.UnitTests
```
Comment (djino97) [inline test/UserService.Mappers.UnitTests/RequestsMappers/UserRequestMapperTests.cs:10]: Add test with Connections equal null

**Review by djino97 — CHANGES_REQUESTED** [review]

File: src/UserService.Models.Db/DbConnection.cs [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        public DbUser User { get; set; }
+    }
+
+    public class DbConnectionConfiguration : IEntityTypeConfiguration<DbConnection>
+    {
+        public void Configure(EntityTypeBuilder<DbConnection> builder)
+        {
+            builder.ToTable("UserConnections");
+            
+            builder.HasKey(pm => new {pm.UserId, pm.Id});
```
[the remark this answers is not in the document]
Comment (borovHokage) [reply src/UserService.Models.Db/DbConnection.cs]: fixed

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace LT.DigitalOffice.UserService.Data.Provider.MsSql.Ef.Migrations
+{
+    [DbContext(typeof(UserServiceDbContext))]
+    [Migration("20210317145607_AddDbConnection")]
+    public class AddUserConnection : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable("UserConnections", table => new
```
[the remark this answers is not in the document]
Comment (borovHokage) [reply src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs]: fixed

File: test/UserService.Validation.UnitTests/UserValidatorTests.cs [hunk 12]
```diff
@@ -224,10 +224,79 @@ public void ShouldPassWhenDataIsValidWithoutMiddleName()
...
+                Connections = new List<UserConnection>()
+                 {
+                    new UserConnection()
+                    {
+                        Type = ConnectionType.Email,
+                        Value = "Ex@mail.ru"
+                    }
+                },
+                Login = "Example"
+            };
```
[the remark this answers is not in the document]
Comment (borovHokage) [reply test/UserService.Validation.UnitTests/UserValidatorTests.cs]: fixed

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs (hunk 8 above)
Comment (borovHokage) [reply src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs]: fixed

File: test/UserService.Mappers.UnitTests/RequestsMappers/UserRequestMapperTests.cs (hunk 9 above)
Comment (borovHokage) [reply test/UserService.Mappers.UnitTests/RequestsMappers/UserRequestMapperTests.cs:10]: fixed

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs (hunk 7 above)
Comment (borovHokage) [reply src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317145607_AddDbConnection.cs]: fixed

File: test/UserService.Mappers.UnitTests/RequestsMappers/UserRequestMapperTests.cs [hunk 13]
```diff
@@ -33,11 +35,19 @@ public void ShouldReturnNewDbUserWhenDataCorrect()
...
+                    { 
+                        Type = ConnectionType.Email,
+                        Value = "Ex@mail.ru"
+                    }
+                }
             };
 
             var result = userRequestMapper.Map(request);
-
+            var conId = result.Connections.FirstOrDefault().Id;
```
[the remark this answers is not in the document]
Comment (borovHokage) [reply test/UserService.Mappers.UnitTests/RequestsMappers/UserRequestMapperTests.cs]: fixed

File: src/UserService.Models.Db/DbConnection.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+        public int Type { get; set; }
+        public string Value { get; set; }
+        public DbUser User { get; set; }
+    }
+
+    public class DbConnectionConfiguration : IEntityTypeConfiguration<DbConnection>
+    {
+        public void Configure(EntityTypeBuilder<DbConnection> builder)
+        {
+            builder.ToTable("UserConnections");
```
[the remark this answers is not in the document]
Comment (borovHokage) [reply src/UserService.Models.Db/DbConnection.cs]: fixed

File: src/UserService.Models.Db/DbConnection.cs [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.UserService.Models.Db
+{
+    public class DbConnection
+    {
```
[the remark this answers is not in the document]
Comment (borovHokage) [reply src/UserService.Models.Db/DbConnection.cs:13]: fixed

File: src/UserService.Models.Dto/UserRequest.cs [hunk 16]
```diff
@@ -14,6 +16,7 @@ public class UserRequest
         public string Password { get; set; }
         public bool IsAdmin { get; set; }
         public bool IsActive { get; set; }
+        public List<UserConnection> Connections { get; set; }
```
Comment (BelehovEgor) [inline src/UserService.Models.Dto/UserRequest.cs:19]: maybe it's better to use IEnumerable instead of List here

Comment (borovHokage) [reply src/UserService.Models.Dto/UserRequest.cs:19]: yes it's possible, but it seemed easier for me using the list(it also works)

Comment (BelehovEgor) [reply src/UserService.Models.Dto/UserRequest.cs:19]: this will also work with IEnumerable because you don't use methods that aren't present in this interface. I think it is not critical to use List, but just in other classes was used IEnumerable

Comment (BelehovEgor) [reply src/UserService.Models.Dto/UserRequest.cs:19]: let's ask @djino97 about this

**Review by BelehovEgor — APPROVED** [review]

File: src/UserService.Models.Db/DbConnection.cs [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    public class DbConnectionConfiguration : IEntityTypeConfiguration<DbConnection>
+    {
+        public void Configure(EntityTypeBuilder<DbConnection> builder)
+        {
+            builder
+                .ToTable(DbConnection.TableName);
+
+            builder.HasKey(id => new {id.UserId, id.Id});
+
+            builder.HasOne(us => us.User)
```
Comment (djino97) [inline src/UserService.Models.Db/DbConnection.cs:32]: Must be `conn`

File: src/UserService.Models.Db/DbConnection.cs [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+                .WithMany(con => con.Connections)
```
Comment (djino97) [inline src/UserService.Models.Db/DbConnection.cs:33]: must be `u`

File: src/UserService.Models.Db/DbConnection.cs [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+                .HasForeignKey(id => id.UserId);
```
Comment (djino97) [inline src/UserService.Models.Db/DbConnection.cs:34]: must be `conn`

**Review by djino97 — CHANGES_REQUESTED** [review]

File: src/UserService.Models.Db/DbUser.cs [hunk 20]
```diff
@@ -23,5 +25,6 @@ public class DbUser
         public DbUserCredentials UserCredentials { get; set; }
         public ICollection<DbUserCertificateFile> CertificatesFilesIds { get; set; }
         public ICollection<DbUserAchievement> AchievementsIds { get; set; }
+        public ICollection<DbConnection> Connections { get; set; }
```
[the remark this answers is not in the document]
Comment (borovHokage) [reply src/UserService.Models.Db/DbUser.cs:30]: fixed

File: src/UserService.Models.Db/DbConnection.cs (hunk 18 above)
Comment (borovHokage) [reply src/UserService.Models.Db/DbConnection.cs]: fixed

File: src/UserService.Models.Db/DbConnection.cs (hunk 19 above)
Comment (borovHokage) [reply src/UserService.Models.Db/DbConnection.cs]: fixed

File: src/UserService.Models.Db/DbConnection.cs [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    public class DbConnectionConfiguration : IEntityTypeConfiguration<DbConnection>
+    {
+        public void Configure(EntityTypeBuilder<DbConnection> builder)
+        {
+            builder
+                .ToTable(DbConnection.TableName);
+
+            builder.HasKey(id => new {id.UserId, id.Id});
```
[the remark this answers is not in the document]
Comment (borovHokage) [reply src/UserService.Models.Db/DbConnection.cs]: fixed

**Review by djino97 — APPROVED** [review]
